### ISSUE TITLE
New Rule: requireRest or disallowArgumentsObject (ES6)

### ISSUE THREAD

**Issue by hzoo** [jscs-dev/node-jscs#1305]
[airbnb/javascript#7.6](https://github.com/airbnb/javascript#7.6)

Never use arguments, opt to use rest syntax ... instead.

> Why? ... is explicit about which arguments you want pulled. Plus rest arguments are a real Array and not Array-like like arguments.

  ```js
// bad
  function concatenateAll() {
    const args = Array.prototype.slice.call(arguments);
    return args.join('');
  }

  // good
  function concatenateAll(...args) {
    return args.join('');
  }
```

**Comment by zxqfox** [comment]
Oh, nice one!

**Comment by markelog** [comment]
Would be a cool rule, but it seems, as complicated one, @hzoo are you up for it?

**Comment by hzoo** [comment]
Hmm my initial thoughts were just to check for the presence of an identifier called `arguments` in a function and error on that? No idea for autofix. And looking at the now I meant `rest` haha.

**Comment by markelog** [comment]
So, basically, something like `disallowIdentifiers` analogues to http://jscs.info/rules.html#disallowkeywords, not sure if that would be enough

**Comment by hzoo** [comment]
I guess we already have `disallowIdentifierNames`

**Comment by markelog** [comment]
Gush, we have 124 rules now. 

It seems we can add it to the airbnb preset

**Comment by hzoo** [comment]
Yeah so this should be covered already - I guess we can close it? The preset is interesting though because that rule would only be turned on for ES6. I don't know if we want to have 2 presets, one for ES6/ES5...

**Comment by markelog** [comment]
We can add new option to the rule, something like 
```js
"disallowIdentifierNames": {
  es6: [],
  es5: []
}
```
it seems strange though

**Comment by hzoo** [comment]
Yeah we could do that for a bunch of rules but it seems messy.

Maybe it makes more sense to make another preset called `airbnb-es6` or something. If it only adds rules, then it could just extend airbnb and add more rules?

**Comment by markelog** [comment]
We have to be careful with amount of presets, not sure if that constitutes the new config

**Comment by mikesherov** [comment]
I think the 

```
"disallowIdentifierNames": {
  es6: [],
  es5: []
}
```

approach is likely best here.

**Comment by hzoo** [comment]
We released [3.0](https://github.com/jscs-dev/node-jscs/releases/tag/v3.0.0) recently! We also wrote a [post](https://medium.com/@markelog/jscs-end-of-the-line-bc9bf0b3fdb2#.oot3wvgpu) about how 3.0 will be our last major release (since we are deprecating the project) and that we are [merging with ESLint](http://eslint.org/blog/2016/04/welcoming-jscs-to-eslint)!

Thus we won't be working on any new features moving forward which includes new rules/options. We will continue to fix bugs and work on CST/better autofixing for current rules in the upcoming months both to help JSCS users and to gain experience that we can hopefully use in ESLint. But otherwise we both teams will be working on getting [ESLint feature parity with JSCS](https://github.com/eslint/eslint/milestones/JSCS%20Compatibility) and just a nice transition for our users.

I'm putting an `orphaned` label to note this issue isn't planned to be worked on, but might be a good contribution as a custom rule (not in JSCS core itself).

I would encourage anyone who wants the rule to try writing it themselves: I would either write a custom rule or submit an issue to ESLint and if accepted a PR. Otherwise a custom rule that won't be in core repo itself.

JSCS rules are also very similar to ESLint rules: here are some resources to get started: 

- [Project-specific Lint Rules with ESLint](https://gist.github.com/jareware/7179093)
- [Creating Custom Rules for ESLint](https://insideops.wordpress.com/2015/12/08/creating-custom-rules-for-eslint)
- [Maintaining JavaScript Code Quality with ESLint ](https://www.paypal-engineering.com/2014/12/12/maintaining-javascript-code-quality-with-eslint/)
